proposals/eligibility_validator.go

```diff
@@ -135,25 +140,18 @@ func (v *Validator) CheckEligibility(ctx context.Context, ballot *types.Ballot,
 // validateReference executed for reference ballots in latest epoch.
 func (v *Validator) validateReference(
 	ballot *types.Ballot,
-	actives []types.ATXID,
-	weight uint64,
+	weight uint64, ref *RefBallotAux,
 ) (*types.EpochData, error) {
 	if ballot.EpochData.Beacon == types.EmptyBeacon {
 		return nil, fmt.Errorf("%w: ref ballot %v", errMissingBeacon, ballot.ID())
 	}
-	if len(actives) == 0 {
+	if ref != nil && ref.TotalWeight == 0 {
 		return nil, fmt.Errorf("%w: ref ballot %v", errEmptyActiveSet, ballot.ID())
 	}
-	totalWeight, used := v.atxsdata.WeightForSet(ballot.Layer.GetEpoch(), actives)
-	for i := range used {
-		if !used[i] {
-			return nil, fmt.Errorf("atx in active set is missing in cache %v", actives[i].ShortString())
-		}
-	}
 	numEligibleSlots, err := GetNumEligibleSlots(
 		weight,
 		minweight.Select(ballot.Layer.GetEpoch(), v.minActiveSetWeight),
-		totalWeight,
+		ref.TotalWeight,
```

Comment:
The `ref != nil` check suggests that the `ref` could be `nil`.  It would then crash on line 154.

Edit: `Handler::checkBallotDataIntegrity()` can return `nil, nil` so it seems that the `ref==nil` case is possible. If so, perhaps add a unit test to cover it.

Reply:
it is not possible to be here if ref is nil, if it is nil we will do validation for secondary ballot. i will check how it can be made better

Reply:
`checkBallotDataIntegrity` can return `nil` for `RefBallotAux` and if it does it would cause a `panic` here.

For me either `checkBallotDataIntegrety` should never return `nil` without an error, or alternatively  - to have a similar behavior as before this change - return `&RefBallotAux{TotalWeight: 0}` (or just not return a pointer).

Reply:
refactored